library/core/src/arch.rs

```diff
@@ -42,3 +42,31 @@ pub macro naked_asm("assembly template", $(operands,)* $(options($(option),*))?)
 pub macro global_asm("assembly template", $(operands,)* $(options($(option),*))?) {
     /* compiler built-in */
 }
+
+/// Compiles to a target-specific software breakpoint instruction or equivalent.
+///
+/// This will typically abort the program. It may result in a core dump, and/or the system logging
+/// debug information. Additional target-specific capabilities may be possible depending on
+/// debuggers or other tooling; in particular, a debugger may be able to resume execution.
+///
+/// If possible, this will produce an instruction sequence that allows a debugger to resume *after*
+/// the breakpoint, rather than resuming *at* the breakpoint; however, the exact behavior is
+/// target-specific and debugger-specific, and not guaranteed.
+///
+/// If the target platform does not have any kind of debug breakpoint instruction, this may compile
+/// to a trapping instruction (e.g. an undefined instruction) instead, or to some other form of
+/// target-specific abort that may or may not support convenient resumption.
+///
+/// The precise behavior and the precise instruction generated are not guaranteed, except that in
+/// normal execution with no debug tooling involved this will not continue executing.
+///
+/// - On x86 targets, this produces an `int3` instruction.
+/// - On aarch64 targets, this produces a `brk #0xf000` instruction.
+// When stabilizing this, update the comment on `core::intrinsics::breakpoint`.
+#[unstable(feature = "breakpoint", issue = "133724")]
+#[inline(always)]
+pub fn breakpoint() {
+    unsafe {
+        core::intrinsics::breakpoint();
```

Comment:
Should update https://github.com/rust-lang/rust/blob/a522d78598415cdd614ccc6d961160f192f64b5c/library/core/src/intrinsics/mod.rs#L1380 and probably just make the intrinsic safe. I see no reason it should be unsafe

Reply:
We could even remove https://github.com/rust-lang/rust/blob/a522d78598415cdd614ccc6d961160f192f64b5c/library/core/src/intrinsics/mod.rs#L1381 and have the body invoke `abort` directly. Allowing codegen backends to just not implement it

Reply:
> Should probably make
> 
> https://github.com/rust-lang/rust/blob/a522d78598415cdd614ccc6d961160f192f64b5c/library/core/src/intrinsics/mod.rs#L1384
> safe. I see no reason it should be unsafe

That seems reasonable to me; will do.

Reply:
> We could even remove
> 
> https://github.com/rust-lang/rust/blob/a522d78598415cdd614ccc6d961160f192f64b5c/library/core/src/intrinsics/mod.rs#L1381
> and have the body invoke `abort` directly. Allowing codegen backends to just not implement it

I don't know if that's the right approach or not. It seems like codegen backends should implement it if at all possible, so that it ends up being a debug trap rather than a non-debug trap.

In any case, I don't think that should be in the same PR; it seems like something that should get decided independently.